llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -20130,6 +20105,135 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+         ++Value) {
+      EVT VT = ValueVTs[Value];
+      MVT RegisterVT =
+          TLI->getRegisterTypeForCallingConv(Context, F.getCallingConv(), VT);
+      unsigned NumRegs =
+          TLI->getNumRegistersForCallingConv(Context, F.getCallingConv(), VT);
+
+      // Skip non-RVV register type
+      if (!RegisterVT.isVector())
+        continue;
+
+      if (RegisterVT.isFixedLengthVector())
+        RegisterVT = TLI->getContainerForFixedLengthVector(RegisterVT);
+
+      RVVArgInfo Info{1, RegisterVT, false};
+
+      while (NumRegs--)
+        RVVArgInfos.push_back(Info);
+    }
+  }
+}
+
+void RVVArgDispatcher::construct(bool IsRet) {
+  const Function &F = MF->getFunction();
+
+  if (IsRet)
+    constructHelper(F.getReturnType());
+  else if (CLI)
+    for (const TargetLowering::ArgListEntry &Arg : CLI->getArgs())
+      constructHelper(Arg.Ty);
+  else if (GISelCLI)
+    for (const CallLowering::ArgInfo &Arg : GISelCLI->OrigArgs)
+      constructHelper(Arg.Ty);
+  else
+    for (const Argument &Arg : F.args())
+      constructHelper(Arg.getType());
+
+  for (auto &Info : RVVArgInfos)
+    if (Info.NF == 1 && Info.VT.getVectorElementType() == MVT::i1) {
+      Info.FirstVMask = true;
+      break;
+    }
+}
+
+void RVVArgDispatcher::allocatePhysReg(unsigned NF, unsigned LMul,
+                                       unsigned StartReg) {
+  assert((StartReg % LMul) == 0 &&
+         "Start register number should be multiple of lmul");
+  const MCPhysReg *VRArrays;
+  switch (LMul) {
+  default:
+    report_fatal_error("Invalid lmul");
+  case 1:
+    VRArrays = ArgVRs;
+    break;
+  case 2:
+    VRArrays = ArgVRM2s;
+    break;
+  case 4:
+    VRArrays = ArgVRM4s;
+    break;
+  case 8:
+    VRArrays = ArgVRM8s;
+    break;
+  }
+
+  for (unsigned i = 0; i < NF; ++i)
+    if (StartReg)
+      AllocatedPhysRegs.push_back(VRArrays[(StartReg - 8) / LMul + i]);
+    else
+      AllocatedPhysRegs.push_back(0);
+}
+
+// This function determines if each RVV argument is passed by register.
+void RVVArgDispatcher::compute() {
+  unsigned ToBeAssigned = RVVArgInfos.size();
+  uint64_t AssignedMap = 0;
+  auto tryAllocate = [&](const RVVArgInfo &ArgInfo) {
+    // Allocate first vector mask argument to V0.
+    if (ArgInfo.FirstVMask) {
+      AllocatedPhysRegs.push_back(RISCV::V0);
+      return;
+    }
+
+    unsigned RegsNeeded = std::max(
+        ArgInfo.VT.getSizeInBits().getKnownMinValue() / RISCV::RVVBitsPerBlock,
+        1UL);
+    unsigned TotalRegsNeeded = ArgInfo.NF * RegsNeeded;
+    for (unsigned StartReg = 0; StartReg + TotalRegsNeeded <= NumArgVRs;
+         StartReg += RegsNeeded) {
+      unsigned Map = ((1 << TotalRegsNeeded) - 1) << StartReg;
+      if ((AssignedMap & Map) == 0) {
+        allocatePhysReg(ArgInfo.NF, RegsNeeded, StartReg + 8);
+        AssignedMap |= Map;
+        return;
+      }
+    }
+
+    allocatePhysReg(ArgInfo.NF, RegsNeeded, 0);
```

Comment:
IIUC, this is the case where there are more regs needed than we have, I.e. if we allocated more than 16 VRs? Do we need a test for this case?

I am not sure I follow why we are allocating physical regs if we have no more physical regs to allocate.

Reply:
If there is no available(follows the spec you mentioned) register that can be allocated, then it is just allocated an invalid MCPhysReg so that in **assign function** it can be passed by reference.